iothub_client/src/iothub_client_diagnostic.c

```diff
@@ -196,3 +196,75 @@ int IoTHubClient_Diagnostic_AddIfNecessary(IOTHUB_DIAGNOSTIC_SETTING_DATA* diagS
 
     return result;
 }
+
+int IoTHubClient_Diagnostic_ParseTwinSettings(IOTHUB_DIAGNOSTIC_SETTING_DATA* diagSetting, const unsigned char* payLoad, char *message) 
+{
+    int result = 0;
+    bool receivedFullTwinForTheFirstTime = true;
+
+    JSON_Value* json = json_parse_string((const char *)payLoad);
+    JSON_Object* jsonObject = json_value_get_object(json);
+    JSON_Object* desiredJsonObject = json_object_get_object(jsonObject, "desired");
+
+    /* Codes_SRS_IOTHUB_DIAGNOSTIC_13_006: [ IoTHubClient_Diagnostic_ParseTwinSettings should retrun nonezero if payLoad is not a valid json string. ]*/
+    if (desiredJsonObject == NULL && jsonObject == NULL)
+    {
+        return __FAILURE__;
+    }
+
+    if (desiredJsonObject == NULL)
+    {
+        desiredJsonObject = jsonObject;
+        receivedFullTwinForTheFirstTime = false;
+    }
+
+    if (!json_object_has_value(desiredJsonObject, DEVICE_TWIN_SAMPLING_RATE_KEY))
+    {        
+        if (receivedFullTwinForTheFirstTime)
+        {
+            /* Codes_SRS_IOTHUB_DIAGNOSTIC_13_007: [ IoTHubClient_Diagnostic_ParseTwinSettings should retrun nonezero if there is no _diag_sample_rate in the twin when received for the first time. ]*/
+            strcat(message, "Property _diag_sample_rate is not exist.");
+            result = __FAILURE__;
+        }
+    }
+    else
+    {
+        double sampling_rate = json_object_get_number(desiredJsonObject, DEVICE_TWIN_SAMPLING_RATE_KEY);
+        if (sampling_rate != 0)
+        {
+            /* Codes_SRS_IOTHUB_DIAGNOSTIC_13_008: [ IoTHubClient_Diagnostic_ParseTwinSettings should retrun nonezero if sampling rate parsed from twin is not between [0,100]. ]*/
+            if (sampling_rate < 0 || sampling_rate > 100)
+            {
+                strcat(message, "Property _diag_sample_rate must between [0, 100].");
+                result = __FAILURE__;
+            }
+            else
+            {
+                diagSetting->diagSamplingPercentage = (int)sampling_rate;
+            }
+        }
+        else
```

Comment:
is there any difference between user set "0" or error status "0"

Reply:
Use a better way to determine json property type